------------------------------------------------------------------------
r47469 | kakuto | 2018-04-28 02:09:56 -0400 (Sat, 28 Apr 2018) | 1 line

dvipdfm-x: Fix a bug which causes a segfault with 1/2/4-bit transparent indexed PNGs (Stefun Br\"uns).

Index: Build/source/texk/dvipdfm-x/configure.ac
===================================================================
--- Build/source/texk/dvipdfm-x/configure.ac	(revision 47468)
+++ Build/source/texk/dvipdfm-x/configure.ac	(revision 47469)
@@ -7,7 +7,7 @@
 dnl   gives unlimited permission to copy and/or distribute it,
 dnl   with or without modifications, as long as this notice is preserved.
 dnl
-AC_INIT([dvipdfm-x (TeX Live)], [20180217], [tex-k@tug.org])
+AC_INIT([dvipdfm-x (TeX Live)], [20180428], [tex-k@tug.org])
 AC_PREREQ([2.65])
 AC_CONFIG_SRCDIR([agl.c])
 AC_CONFIG_AUX_DIR([../../build-aux])
Index: Build/source/texk/dvipdfm-x/pngimage.c
===================================================================
--- Build/source/texk/dvipdfm-x/pngimage.c	(revision 47468)
+++ Build/source/texk/dvipdfm-x/pngimage.c	(revision 47469)
@@ -964,6 +964,7 @@
   png_bytep   trans;
   int         num_trans;
   png_uint_32 i;
+  png_byte    bpc, mask, shift;
 
   if (!png_get_valid(png_ptr, info_ptr, PNG_INFO_tRNS) ||
       !png_get_tRNS(png_ptr, info_ptr, &trans, &num_trans, NULL)) {
@@ -970,6 +971,9 @@
     WARN("%s: PNG does not have valid tRNS chunk but tRNS is requested.", PNG_DEBUG_STR);
     return NULL;
   }
+  bpc   = png_get_bit_depth(png_ptr, info_ptr);
+  mask  = 0xff >> (8 - bpc);
+  shift = 8 - bpc;
 
   smask = pdf_new_stream(STREAM_COMPRESS);
   dict  = pdf_stream_dict(smask);
@@ -981,7 +985,8 @@
   pdf_add_dict(dict, pdf_new_name("ColorSpace"), pdf_new_name("DeviceGray"));
   pdf_add_dict(dict, pdf_new_name("BitsPerComponent"), pdf_new_number(8));
   for (i = 0; i < width*height; i++) {
-    png_byte idx = image_data_ptr[i];
+    /* data is packed for 1/2/4 bpc formats, msb first */
+    png_byte idx = (image_data_ptr[bpc * i / 8] >> (shift - bpc * i % 8)) & mask;
     smask_data_ptr[i] = (idx < num_trans) ? trans[idx] : 0xff;
   }
   pdf_add_stream(smask, (char *)smask_data_ptr, width*height);

------------------------------------------------------------------------
